src/decoder/builder.rs

```diff
@@ -0,0 +1,307 @@
+    pub(crate) hint: Option<String>,
+
+    /// An MIME type hint for the decoder about the format of the stream.
+    /// When known, this can help the decoder to select the correct demuxer.
+    pub(crate) mime_type: Option<String>,
+
+    /// Whether the decoder should report as seekable.
+    pub(crate) is_seekable: bool,
+}
+
+impl Default for Settings {
+    fn default() -> Self {
+        Self {
+            byte_len: None,
+            coarse_seek: false,
+            gapless: true,
+            hint: None,
+            mime_type: None,
+            is_seekable: false,
+        }
+    }
+}
+
+/// Builder for configuring and creating a decoder.
+///
+/// This provides a flexible way to configure decoder settings before creating
+/// the actual decoder instance.
+///
+/// # Examples
+///
+/// ```no_run
+/// use std::fs::File;
+/// use rodio::decoder::DecoderBuilder;
+///
+/// let file = File::open("audio.mp3").unwrap();
+/// let decoder = DecoderBuilder::new()
+///     .with_data(file)
+///     .with_hint("mp3")
+///     .with_gapless(true)
+///     .build()
+///     .unwrap();
+/// ```
+#[derive(Clone, Debug)]
+pub struct DecoderBuilder<R> {
+    /// The input data source to decode.
+    data: Option<R>,
+    /// Configuration settings for the decoder.
+    settings: Settings,
+}
+
+impl<R> Default for DecoderBuilder<R> {
+    fn default() -> Self {
+        Self {
+            data: None,
+            settings: Settings::default(),
+        }
+    }
+}
+
+impl<R: Read + Seek + Send + Sync + 'static> DecoderBuilder<R> {
+    /// Creates a new decoder builder with default settings.
+    ///
+    /// # Examples
+    /// ```no_run
+    /// use std::fs::File;
+    /// use rodio::decoder::DecoderBuilder;
+    ///
+    /// let file = File::open("audio.mp3").unwrap();
+    /// let decoder = DecoderBuilder::new()
+    ///     .with_data(file)
+    ///     .build()
+    ///     .unwrap();
+    /// ```
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Sets the input data source to decode.
+    pub fn with_data(mut self, data: R) -> Self {
+        self.data = Some(data);
+        self
+    }
+
+    /// Sets the byte length of the stream.
+    /// This is required for:
+    /// - Reliable seeking operations
+    /// - Duration calculations in formats that lack timing information (e.g. MP3, Vorbis)
+    ///
+    /// The byte length should typically be obtained from file metadata:
+    /// ```no_run
+    /// use std::fs::File;
+    /// use rodio::Decoder;
+    ///
+    /// let file = File::open("audio.mp3").unwrap();
+    /// let len = file.metadata().unwrap().len();
+    /// let decoder = Decoder::builder()
+    ///     .with_data(file)
+    ///     .with_byte_len(len)
+    ///     .build()
+    ///     .unwrap();
+    /// ```
+    ///
+    /// Alternatively, it can be obtained by seeking to the end of the stream.
+    pub fn with_byte_len(mut self, byte_len: u64) -> Self {
```

Comment:
might be worth to mention what happens if I set an incorrect byte_len. What does byte_len smaller then actual len do? Same for the inverse.

Reply:
I'll add something and keep it short. Basically it's just wrong and I propose we shy away from making any promises about what happens then. Currently, Symphonia is likely to report incorrect `total_duration` and seek sub-optimally or to unexpected positions.

Reply:
> I propose we shy away from making any promises about what happens then.

agreed. In that case we should note that